### PR TITLE
Represent metadata as a ByteBuffer

### PR DESCRIPTION
**Pull request by segabriel**
Closes https://github.com/scalecube/scalecube-cluster/issues/30

### CODE REVIEW

File: cluster/src/test/java/io/scalecube/cluster/metadata/SimpleMapMetadataCodec.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
+package io.scalecube.cluster.metadata;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.nio.ByteBuffer;
+import java.util.Map;
+import reactor.core.Exceptions;
+
+public class SimpleMapMetadataCodec implements MetadataCodec<Map<String, String>> {
```
Comment (ronenhamias) [inline cluster/src/test/java/io/scalecube/cluster/metadata/SimpleMapMetadataCodec.java]: should be  MetadataCodec<Map<String, ByteBuffer>> and not sting IMHO

Comment (segabriel) [reply cluster/src/test/java/io/scalecube/cluster/metadata/SimpleMapMetadataCodec.java]: no, it's a client type

File: cluster/src/main/java/io/scalecube/cluster/ClusterConfig.java [hunk 2]
```diff
@@ -268,13 +278,17 @@ public String toString() {
...
+        : "metadata@" + Integer.toHexString(metadata.hashCode()) + "[" + metadata.remaining() + "]";
   }
 
   public static final class Builder {
 
+    public static final ByteBuffer EMPTY = ByteBuffer.wrap(new byte[0]);
+
     private List<Address> seedMembers = Collections.emptyList();
-    private Map<String, String> metadata = new HashMap<>();
+    private ByteBuffer metadata = EMPTY;
```
Comment (artem-v) [inline cluster/src/main/java/io/scalecube/cluster/ClusterConfig.java]: IMHO `ClusterConfig`'s (further `CF`)  `metadata`  has nothing to do with ByteBuffer, it's  `MetadataCodec`  must be concerned about encoding CF.metadata to advertise it to the cluster. BTW, for local CF metadata only process of converting `Object` -> `Bytes` is relevant.  For a part where we getting remote metadata then decoding process (`Bytes` -> `Object`  )  is taking place.

File: cluster/src/main/java/io/scalecube/cluster/ClusterConfig.java [hunk 3]
```diff
@@ -290,24 +304,15 @@ private String metadataAsString() {
...
-    }
-
-    public Builder addMetadata(String key, String value) {
-      this.metadata.put(key, value);
-      return this;
-    }
-
-    public Builder addMetadata(Map<String, String> metadata) {
-      this.metadata.putAll(metadata);
+    public Builder metadata(ByteBuffer metadata) {
```
Comment (artem-v) [inline cluster/src/main/java/io/scalecube/cluster/ClusterConfig.java]: For local metadata this has to be an `Object`,  not a ByteBuffer.

File: cluster/src/main/java/io/scalecube/cluster/metadata/MetadataCodec.java [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package io.scalecube.cluster.metadata;
+
+import java.nio.ByteBuffer;
+
+public interface MetadataCodec<T> {
+
+  T deserialize(ByteBuffer buffer);
+
+  ByteBuffer serialize(T metadata);
+}
```
Comment (artem-v) [inline cluster/src/main/java/io/scalecube/cluster/metadata/MetadataCodec.java]: Let's split this into two interfaces. 
`MetadataEncoder.encode() : LOCAL_T -> BB `
and  
`MetadataDecoder.decode() : BB -> REMOTE_T`

**Pull request merged**